### PR TITLE
Csra 429 submission api

### PR DESCRIPTION
**Pull request by glenjamin**
This adds the Joi validation, which should bring this up to a production-ready state.

### CODE REVIEW

File: test/unit/server/services/assessment.spec.js [hunk 1]
```diff
@@ -60,6 +60,110 @@ describe('assessment service', () => {
...
+      it('returns validation error', () => {
+        expect(error).to.be.an('error');
+        expect(error).to.have.property('type', 'validation');
+      });
+      it('doesnt talk to the database', () => {
+        expect(fakeDB.insert.callCount).to.eql(0);
+      });
+    });
+    describe('rules', () => {
+      function allow(data, label) {
```
Comment (retroredge) [inline test/unit/server/services/assessment.spec.js]: "allows" would read better where this function is used

File: test/unit/server/services/assessment.spec.js [hunk 2]
```diff
@@ -60,6 +60,110 @@ describe('assessment service', () => {
...
+        const payload = Object.assign({}, validRiskAssessment, data);
+        it(`allows ${label || JSON.stringify(data)}`, () =>
+          expect(assessment.record(payload)).to.be.fulfilled);
+      }
+      function deny(data, label) {
```
Comment (retroredge) [inline test/unit/server/services/assessment.spec.js]: "doesNotAllow" would read better where this function is used

File: server/services/assessment.js [hunk 3]
```diff
@@ -1,9 +1,47 @@
...
+  const validated = Joi.validate(rawAssessment, schema, {
+    abortEarly: false,
+    presence: 'required',
+  });
+
+  if (validated.error) {
+    const err = new Error(`Validation failed: ${validated.error.message}`);
+    err.type = 'validation';
+    err.details = validated.error.details;
+    return Promise.reject(err);
```
Comment (retroredge) [inline server/services/assessment.js]: Is asynchronous code needed here?

Comment (glenjamin) [inline server/services/assessment.js]: Yes, but only to preserve the signature - it always returns a Promise.

File: test/mocha.opts [hunk 4]
```diff
@@ -1,4 +1,3 @@
 --require ./test/.setup.js
 --reporter spec
 --colors
---bail
```
Comment (romidane) [inline test/mocha.opts]: Can you add this back :)

File: test/unit/server/services/assessment.spec.js [hunk 5]
```diff
@@ -60,6 +60,110 @@ describe('assessment service', () => {
     ],
   };
 
+  describe('validation', () => {
+    describe('general validation stuff', () => {
```
Comment (romidane) [inline test/unit/server/services/assessment.spec.js]: can you change the describe to a context?

**Pull request merged**